internal/server/apimigrator.go

```diff
@@ -0,0 +1,157 @@
+package server
+
+import (
+	"bufio"
+	"encoding/json"
+	"net"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+)
+
+type apiMigration struct {
+	method          string
+	path            string
+	version         string
+	requestRewrite  func(c *gin.Context)
+	responseRewrite func(c *gin.Context)
+}
+
+var migrations = []apiMigration{}
+
+func addResponseRewrite[newResp any, oldResp any](method, path, version string, f func(newResp) oldResp) {
+	migrations = append(migrations, apiMigration{
+		method:  method,
+		path:    path,
+		version: version,
+		responseRewrite: func(c *gin.Context) {
+			reqVer := c.Request.Header.Get("X-Infra-Version")
+			if reqVer == "" || NewVersion(reqVer).GreaterThanStr(version) {
+				c.Next()
+				return
+			}
+
+			w := &responseWriter{ResponseWriter: c.Writer}
+			c.Writer = w
+
+			c.Next()
+
+			newRespObj := new(newResp)
+			err := json.Unmarshal(w.body, newRespObj)
+			if err != nil {
+				panic(err)
+			}
+
+			oldRespObj := f(*newRespObj)
+
+			b, err := json.Marshal(oldRespObj)
+			if err != nil {
+				panic(err)
+			}
+
+			w.body = b
+			w.Flush()
+
+			if w.flushErr != nil {
+				panic(w.flushErr)
+			}
+		},
+	})
+}
+
+type responseWriter struct {
```

Comment:
Having to wrap the response writer seems like it's going to be the biggest maintenance cost of this implementation.  

I wonder if we could avoid this somehow. Our API endpoints currently return types. If the response migration could accept types and translate those types (before encoding to JSON) that might remove the need to wrap ResponseWriter? I don't know how we do that in a generic way, but it seems easy enough to do by wrapping the function with another one.

Reply:
could be doable, but sort of breaks the middleware approach. Let's look at this as an optimization later.